### PR TITLE
Teuchos: Fix LARND with Kokkos::complex<double>

### LINKED ISSUE #12385: Teuchos:  clang reporting improper cast in Teuchos_LAPACK.cpp
## Bug Report
@bartlettroscoe not really sure who owns this one but the error is happening when compiling _Teuchos\_LAPACK.cpp_ so you're the "lucky" winner.

### Description
As of last night I'm getting the following error when compiling _Teuchos\_LAPACK.cpp_ from the _develop_ branch:
```
/Trilinos/packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:2096:49: error: implicit conversion from '_Complex double' to 'const double' is not permitted in C++
    return static_cast<Kokkos::complex<double>>(z);
           ~~~~~~~~~~~                          ^
1 error generated.
```

### PR DESCRIPTION
**Pull request by vqd8a**
@trilinos/teuchos

## Motivation

A small PR to fix #12385.

### CODE REVIEW

**Review by hkthorn — APPROVED** [review]
LGTM

File: packages/teuchos/numerics/src/Teuchos_LAPACK.cpp [hunk 1]
```diff
@@ -2093,7 +2093,7 @@ namespace Teuchos
   Kokkos::complex<double> LAPACK<int, Kokkos::complex<double> >::LARND( const int& idist, int* seed ) const
   {
     double _Complex z = ZLARND_F77(&idist, seed);
-    return static_cast<Kokkos::complex<double>>(z);
+    return reinterpret_cast<Kokkos::complex<double>&>(z);
```
[the remark this answers is not in the document]
Comment (hkthorn) [reply packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:2096]: The `reinterpret_cast<>()` was already added to the std::complex<double> and std::complex<float> specializations in this commit: https://github.com/trilinos/Trilinos/commit/900b9651845be654f6ee0f92c766bd67c3f84066

**Review by hkthorn — APPROVED** [review]
Given the precedent, I can accept this.

File: packages/teuchos/numerics/src/Teuchos_LAPACK.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:2096]: @hkthorn, are we concerned that there is no Trilinos Framework-run PR or Nightly build that is triggering these errors?

Also, why has this change not been make to all of the LAPACK and BLAS complex functions?

[the remark this answers is not in the document]
Comment (hkthorn) [reply packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:2096]: So this is only an issue when a complex value is returned by the method, which is only an issue for BLAS::DOT and LAPACK::LARND.  Both of those methods are using the reinterpret_cast after the above commit.

[the remark this answers is not in the document]
Comment (hkthorn) [reply packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:2096]: Also, yes, it is disturbing that this was not caught by PR testing.  :)

**Comment by bartlettroscoe** [comment]
> Also, yes, it is disturbing that this was not caught by PR testing. :)

@hkthorn and @glhenni, would one of you mind posting a [Trilinos Help Desk ticket](https://sems-atlassian-son.sandia.gov/jira/servicedesk/customer/portal/7) to cover this complex code in Trilinos PR testing (or at least nightly testing)?  It will be made a higher priority if it comes from an important Trilinos customer.

**Comment by hkthorn** [comment]
Failure appears unrelated.  Trying retest.

**Pull request merged**